src/server/status_server.rs

```diff
@@ -210,6 +212,142 @@ impl StatusServer {
+            (Regex::new(r"^snap-sender\d*$").unwrap(), "snap-sender"),
+            (Regex::new(r"^apply-\d*$").unwrap(), "apply"),
+            (Regex::new(r"^future-poller-\d*$").unwrap(), "future-poller"),
+        ];
+
+        move |frames| {
+            for (regex, name) in thread_rename.iter() {
+                if regex.is_match(&frames.thread_name) {
+                    frames.thread_name = name.to_string();
+                }
+            }
+        }
+    }
+
+    pub fn dump_rsprof(
+        seconds: u64,
+        frequency: i32,
+    ) -> Box<dyn Future<Item = pprof::Report, Error = pprof::Error> + Send> {
+        match pprof::ProfilerGuard::new(frequency) {
+            Ok(guard) => {
+                info!(
+                    "start profiling {} seconds with frequency {} /s",
+                    seconds, frequency
+                );
+
+                let timer = GLOBAL_TIMER_HANDLE.clone();
+                Box::new(
+                    timer
+                        .delay(std::time::Instant::now() + std::time::Duration::from_secs(seconds))
+                        .then(
+                            move |_| -> Box<
+                                dyn Future<Item = pprof::Report, Error = pprof::Error> + Send,
+                            > {
+                                let _ = guard;
+                                Box::new(
+                                    match guard
+                                        .report()
+                                        .frames_post_processor(Self::frames_post_processor())
+                                        .build()
+                                    {
+                                        Ok(report) => ok(report),
+                                        Err(e) => err(e),
+                                    },
+                                )
+                            },
+                        ),
+                )
+            }
+            Err(e) => Box::new(err(e)),
+        }
+    }
+
+    pub fn dump_rsperf_to_resp(
+        req: Request<Body>,
+    ) -> Box<dyn Future<Item = Response<Body>, Error = hyper::Error> + Send> {
+        let query = match req.uri().query() {
+            Some(query) => query,
+            None => {
+                let response = Response::builder()
+                    .status(StatusCode::BAD_REQUEST)
+                    .body(Body::empty())
+                    .unwrap();
+                return Box::new(ok(response));
+            }
+        };
+        let query_pairs: HashMap<_, _> = url::form_urlencoded::parse(query.as_bytes()).collect();
+        let seconds: u64 = match query_pairs.get("seconds") {
+            Some(val) => match val.parse() {
+                Ok(val) => val,
+                Err(_) => {
+                    let response = Response::builder()
+                        .status(StatusCode::BAD_REQUEST)
+                        .body(Body::empty())
```

Comment:
Prefer to use the error message as the response body to make diagnose easily (many errors will return `BAD_REQUEST` and the user cannot figure out which one is the cause of `BAD_REQUEST`).